### PR TITLE
Epoch single pass

### PR DESCRIPTION
**Pull request by dapplion**
Cherry pick epoch single pass feature from https://github.com/sigp/lighthouse/pull/3206. For more details on the feature:
- https://github.com/sigp/lighthouse/pull/4483

### TODOs

https://github.com/dapplion/lighthouse/blob/23e10c6d59adf81e49b331bfc65e85acad5ebbe5/consensus/state_processing/src/common/update_progressive_balances_cache.rs#L29

https://github.com/dapplion/lighthouse/blob/23e10c6d59adf81e49b331bfc65e85acad5ebbe5/consensus/state_processing/src/per_epoch_processing/altair/participation_cache.rs#L475-L476

- [x] Fixed by dropping the entire ParticipationCache https://github.com/dapplion/lighthouse/pull/16

https://github.com/dapplion/lighthouse/blob/23e10c6d59adf81e49b331bfc65e85acad5ebbe5/consensus/state_processing/src/per_block_processing/altair/sync_committee.rs#L55

- [x] Fix for the optimization to remain spec compliant https://github.com/ChainSafe/lodestar/blame/unstable/packages/state-transition/src/block/processSyncCommittee.ts. Fixed with https://github.com/dapplion/lighthouse/pull/15

https://github.com/dapplion/lighthouse/blob/23e10c6d59adf81e49b331bfc65e85acad5ebbe5/consensus/state_processing/src/per_epoch_processing/epoch_processing_summary.rs#L146

- [x] Just delete, not used, only for metrics. Fixed https://github.com/dapplion/lighthouse/pull/13/commits/f420033e72ae316e18d79564734e97c7ce694568

https://github.com/dapplion/lighthouse/blob/23e10c6d59adf81e49b331bfc65e85acad5ebbe5/consensus/types/src/chain_spec.rs#L323

- [x] Fixed https://github.com/dapplion/lighthouse/pull/13/commits/90ee50fbf37b1fcf543d3528786dfdb77e4808fe

### CODE REVIEW

File: consensus/state_processing/src/all_caches.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                .progressive_balances_cache()
+                .is_initialized_at(current_epoch)
+            && self.pubkey_cache().len() == self.validators().len()
+            && self.exit_cache().check_initialized().is_ok()
+            && self.slashings_cache_is_initialized()
+            && self
+                .epoch_cache()
+                .check_validity::<E>(current_epoch, epoch_cache_decision_block_root)
+                .is_ok()
+    }
```
Comment (dapplion) [inline consensus/state_processing/src/all_caches.rs:53]: Should this include the tree_hash_cache?

File: consensus/state_processing/src/per_epoch_processing/single_pass.rs [hunk 2]
```diff
@@ -0,0 +1,628 @@
...
+    ) {
+        let validator = validators_iter
+            .next()
+            .ok_or(BeaconStateError::UnknownValidator(index))?;
+        let balance = balances_iter
+            .next()
+            .ok_or(BeaconStateError::UnknownValidator(index))?;
+        let inactivity_score = inactivity_scores_iter
+            .next()
+            .ok_or(BeaconStateError::UnknownValidator(index))?;
```
Comment (dapplion) [inline consensus/state_processing/src/per_epoch_processing/single_pass.rs:195]: Should we keep this syntax or zip all the lists? I would keep as-is to reduce diff latter

File: consensus/types/src/beacon_state/pubkey_cache.rs [hunk 3]
```diff
@@ -4,6 +4,7 @@ use std::collections::HashMap;
 
 type ValidatorIndex = usize;
 
+#[allow(clippy::len_without_is_empty)]
```
Comment (dapplion) [inline consensus/types/src/beacon_state/pubkey_cache.rs:7]: Why is clippy complaining in this PR and not in unstable?

File: beacon_node/operation_pool/src/attestation.rs [hunk 4]
```diff
@@ -99,8 +98,11 @@ impl<'a, T: EthSpec> AttMaxCover<'a, T> {
 
                 let mut proposer_reward_numerator = 0;
 
+                // FIXME(sproul): store base_reward in reward cache
```
Comment (michaelsproul) [inline beacon_node/operation_pool/src/attestation.rs]: I think we can resolve this by using the base reward from the `EpochCache`. Trying that now

File: beacon_node/beacon_chain/src/attestation_rewards.rs [hunk 5]
```diff
@@ -199,24 +195,41 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
             Self::validators_ids_to_indices(&mut state, validators)?
         };
 
-        for validator_index in &validators {
-            let eligible = state.is_eligible_validator(previous_epoch, *validator_index)?;
+        for &validator_index in &validators {
+            // Return 0s for unknown/inactive validator indices. This is a bit different from stable
+            // where we error for unknown pubkeys.
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/attestation_rewards.rs:197]: We should decide on whether these semantics are OK, and amend this comment accordingly.

File: consensus/state_processing/src/per_epoch_processing/single_pass.rs (hunk 2 above)
Comment (michaelsproul) [reply consensus/state_processing/src/per_epoch_processing/single_pass.rs:195]: Agree keeping as-is is good

**Comment by dapplion** [comment]
- Closing for https://github.com/sigp/lighthouse/pull/5279

**Pull request closed without merging**